ulscrape/spiders/uradni_list.py

```diff
@@ -15,39 +15,44 @@ def __init__(self, years=None, *args, **kwargs):
         """ Initialize original Scrapy spider. Then load arguments passed via CLI. """
         super(UradniListSpider, self).__init__(*args, **kwargs)
 
-        if years is not None or years is not '':
+        if years is not None and years is not '':
             self.initial_years = [int(y) for y in years.split(",") if y != '']
 
     def start_requests(self):
        return [scrapy.http.FormRequest(
            url=self.search_url, formdata={'year': str(year)}
        ) for year in self.search_years()]
 
-
     def parse(self, response):
+        if '&' not in response.request.body.decode(encoding='utf-8'):
+            return self.parse_archive_index_page(response)
+        elif '&' in response.request.body.decode(encoding='utf-8'):
+            return self.parse_archive_page(response)
+        else:
+            raise Exception('No match for parsing the UL index or archive page, strange.')
+
+    def parse_archive_index_page(self, response):
```

Comment:
This method needs some more love. This is what you can do,...
- [ ] Your regex needs some more love. Improve it [on regex101](https://regex101.com/) - there is lovely cheet sheet on the bottom - use it! 
a. You don't need named groups if you are looking for just one match.
b. If you are looking for numbers in text please use `\d` or if you are looking for more than one number `\d+`
c. Simple regex is always faster - remember that!
d. Following example uses `global` flag `g`,...
```regex
/val\(\'(\d+)\'\)/g
```
- [ ] [re.compile()](https://docs.python.org/3/library/re.html?highlight=re.compile#re.compile) is boring... use [re.findall()](https://docs.python.org/3/library/re.html?highlight=re.findall#re.findall)
```python
pages = re.findall(r'val\(\'(\d+)\'\)', response.text)
```
- [ ] After you get page number matches, just loop over them, convert them to number with [int()](https://docs.python.org/3/library/functions.html?highlight=max#int) and find largest number with [max()](https://docs.python.org/3/library/functions.html?highlight=max#max),...
a. By now you already know [in-line list comprehension](http://blog.teamtreehouse.com/python-single-line-loops), so you can do it in single line.
```python
page_numbers = [int(page) for page in pages]
largest_page = max(page_numbers)
```
- [ ] You have problem that you constantly need to check for year in response. That takes time and is messy. Scrapy uses [request.meta or response.meta](https://doc.scrapy.org/en/latest/topics/request-response.html#scrapy.http.Request.meta) for that... This is what you can do:
a. Add `meta.year` to `start_requests()`
```python
def start_requests(self):
  return [scrapy.http.FormRequest(
    url=self.search_url,
    formdata={'year': str(year)},
    meta={'year': year}
  ) for year in self.search_years()]
```
b. Then you can leverage `response.meta['year']` in your parsers...
```python
def parse_index_page(self, response):
  year = response.meta['year']
  max_page = max([int(page) for page in re.findall(r'val\(\'(\d+)\'\)', response.text)])
  print('year = {} max_page = {}'.format(year, max_page))
```
- [ ] Scrapy has method [response.text](https://doc.scrapy.org/en/latest/topics/request-response.html#scrapy.http.TextResponse.text) that gives you back body as string... So you don't have to do `response.body.decode(encoding='utf-8')`...
- [ ] Please use proper imports! Like this:
```python
from scrapy import Request, FormRequest
```
- [ ] You started using list, and range! That is awesome!!! I'm very very happy to see that! Not use that logic with generators and you get this:
a. Don't use range and than list over it... It destroys the purpose of range, don't you think? 😄 
```python
return [FormRequest(
    url=self.search_url,
    formdata={'year': str(year), 'page': str(page)},
    meta={'year': year, 'page': page}
) for page in range(1, max_page + 1)]
```